Autoformat the Node SDK

RobbieMcKinstry opened the pull request:
<!--- 
Thanks so much for your contribution! If this is your first time contributing, please ensure that you have read the [CONTRIBUTING](https://github.com/pulumi/pulumi/blob/master/CONTRIBUTING.md) documentation.
-->

# Description

<!--- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. -->

This PR adds Rome as an autoformatter for the NodeJS SDK. It applies the autoformat rules to the codebase.
In the first commit, you can see the changes I've made to the toolchain to install Rome and include it in
CI. In the second commit, I've run the formatter over the codebase.

Fixes https://github.com/pulumi/pulumi/issues/12781

## Checklist

Not applicable. This is not a user-facing feature, so no changelog entry is required.
Additionally, because this is a pure stlye refactor, I've not added any unit tests to
verify the feature.

<!--- Please provide details if the checkbox below is to be left unchecked. -->
- [ ] I have added tests that prove my fix is effective or that my feature works
<!--- 
User-facing changes require a CHANGELOG entry.
-->
- [ ] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change
<!--
If the change(s) in this PR is a modification of an existing call to the Pulumi Cloud,
then the service should honor older versions of the CLI where this change would not exist.
You must then bump the API version in /pkg/backend/httpstate/client/api.go, as well as add
it to the service.
-->
- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version
  <!-- @Pulumi employees: If yes, you must submit corresponding changes in the service repo. -->

RobbieMcKinstry commented:
Current dependencies on/for this PR:
* master
  * **PR #12789** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/12789" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/pulumi/pulumi/12789?utm_source=stack-comment).

RobbieMcKinstry commented on sdk/nodejs/.eslintrc.js:
```diff
@@ -1,84 +1,68 @@
-                "accessibility": "explicit"
-            }
-        ],
-        "@typescript-eslint/indent": [
-            "error",
-            4,
-            {
-                "FunctionDeclaration": {
-                    "parameters": "first"
-                },
-                "FunctionExpression": {
-                    "parameters": "first"
-                }
-            }
```
This is probably the most "controversial" change. ESLint is doing double duty for us, policing both linting _and_ formatting. This change removes ESLint's "indentation" rule because it's incompatible with Rome's rules on some minor edge cases: switch statement cases indentation, nested ternary operators, etc. Otherwise, the two tools are compatible.

abhinav commented on sdk/nodejs/Makefile:
```diff
@@ -27,6 +27,12 @@ ensure:: yarn.ensure node.ensure .ensure.phony
+	yarn rome format --write .
+
+format_ci:: ensure
+	yarn rome ci .
```
Just to confirm: this will still verify that code that is poorly formatted fails in CI, right? That's my read of this comment and the Rome docs, but I want to confirm.

RobbieMcKinstry replied:
Yes, `rome ci` sets the status code to non-zero if there's formatting violations, which fails CI. You can run it locally too; it doesn't run exclusively in CI environments.

abhinav commented on sdk/nodejs/yarn.lock:
```diff
@@ -2645,6 +2675,18 @@ rimraf@^3.0.0, rimraf@^3.0.2:
   dependencies:
     glob "^7.1.3"
 
+rome@^12.0.0:
```
Shouldn't there be an accompanying package.json change?

RobbieMcKinstry replied:
Yes, [it's here](https://github.com/pulumi/pulumi/pull/12789/commits/2e12a4cb4b04366c635483a488409335f0b1e09d#diff-0d53ce0d334b84822a2dbdf673d0456efbacb59a76f72aabbfb1b7f8203a58f7R55).

abhinav commented on sdk/nodejs/rome.json:
```diff
@@ -0,0 +1,18 @@
+{
```
Rome's default lineWidth is too aggressive (80 chars). https://docs.rome.tools/configuration/#formatterlinewidth

While I prefer 80 in general, most of our code uses 120 characters. We should probably use that here too.

abhinav commented on sdk/nodejs/rome.json:
```diff
@@ -0,0 +1,20 @@
+{
+    "$schema": "./node_modules/rome/configuration_schema.json",
+    "organizeImports": {
+        "enabled": false
+    },
```
I don't have a strong preference on this, but I'm curious why we want to turn this off.

RobbieMcKinstry replied:
Three reasons that aren't of consequence:
1. That was the default.
2. I didn't want to make more changes than necessary in the initial land.
3. ESLint provides some guidelines for import order already. So we should already have import org enforced at the linter level (not the auto-fmt level, which would be better). I don't actually know if they are compatible; if not, we should enable them in Rome and disable them in ESLint so it "just works" with auto-fmt.

WDYT? Should we land this first and then toy with it, turn it on now, or defer to ESLint?

abhinav replied:
Those are valid reasons. As I said, i didn't have a preference. I didn't realize that this was the default.

RobbieMcKinstry commented:
@abhinav I've addressed your feedback around the 120 character limit. Everything should now be formatted as if I had originally set the character limit to 120, instead of 80.

RobbieMcKinstry commented:
GH isn't showing it from the Conversations tab, (because it's a commit comment, not a PR comment?) [but I replied to this](https://github.com/pulumi/pulumi/pull/12789#discussion_r1180832448).

abhinav reviewed (approved):
Great! Thanks for making all the updates.

RobbieMcKinstry commented:
My pleasure! Thanks for being open to new tech. I appreciate that my team is always learning <3

RobbieMcKinstry commented:
bors r+
